pkg/api/dirs.go

```diff
@@ -31,26 +30,46 @@ import (
 )
 
 const (
-	contentTypeHeader = "Content-Type"
-	contentTypeTar    = "application/x-tar"
+	contentTypeTar = "application/x-tar"
 )
 
 const (
 	manifestRootPath                      = "/"
 	manifestWebsiteIndexDocumentSuffixKey = "website-index-document"
 	manifestWebsiteErrorDocumentPathKey   = "website-error-document"
+	manifestEntryMetadataContentTypeKey   = "Content-Type"
+	manifestEntryMetadataFilenameKey      = "Filename"
 )
 
 // dirUploadHandler uploads a directory supplied as a tar in an HTTP request
 func (s *server) dirUploadHandler(w http.ResponseWriter, r *http.Request) {
 	logger := tracing.NewLoggerWithTraceID(r.Context(), s.logger)
-	err := validateRequest(r)
-	if err != nil {
-		logger.Errorf("dir upload, validate request")
-		logger.Debugf("dir upload, validate request err: %v", err)
+	if r.Body == http.NoBody {
+		logger.Error("dir upload, request has no body")
 		jsonhttp.BadRequest(w, "could not validate request")
 		return
 	}
+	contentType := r.Header.Get(contentTypeHeader)
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		logger.Errorf("dir upload, invalid content-type")
+		logger.Debugf("dir upload, invalid content-type err: %v", err)
+		jsonhttp.BadRequest(w, "invalid content-type")
+		return
+	}
+
+	var dReader dirReader
+	switch mediaType {
+	case contentTypeTar:
+		dReader = &tarReader{r: tar.NewReader(r.Body), logger: s.logger}
+	case multiPartFormData:
+		dReader = &multipartReader{r: multipart.NewReader(r.Body, params["boundary"]), logger: s.logger}
+	default:
+		logger.Error("dir upload, invalid content-type for directory upload")
```

Comment:
prefix log message with package context:
`api: dir upload: invalid...`

Reply:
I meant to change this. I was thinking of using the prefix "bzz upload dir:", this would follow the one I have used for file upload which is "bzz upload file:"

In order to use the "api:" prefix we should use a prefixed logger in the server itself. I think api is not prefixed in any handler.